pkg/smartcontract/manifest/permission.go

```diff
@@ -0,0 +1,174 @@
+package manifest
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// PermissionType represents permission type.
+type PermissionType uint8
+
+const (
+	// PermissionWildcard allows everything.
+	PermissionWildcard PermissionType = 0
+	// PermissionHash restricts called contracts based on hash.
+	PermissionHash PermissionType = 1
+	// PermissionGroup restricts called contracts based on public key.
+	PermissionGroup PermissionType = 2
+)
+
+// PermissionDesc is a permission descriptor.
+type PermissionDesc struct {
+	Type  PermissionType
+	Value interface{}
```

Comment:
It will become harder to write unmarshaler, as we will need to implement it for every structure which contains an interface.
Or one can define an unmarshaler for a struct containing an interface.
Anyway I don't see how this simplifies anything, given that `Type` + `Value` is the way it is represented in JSON.